Fixes visit/2 issue where urls have too many slashes

aaronrenner opened the pull request:
I noticed in #300 that `visit/2` is generating urls like http://localhost:54579//logs.html that have too many slashes. This occurs when the base_url has a trailing slash and the requested path has a leading slash. This change adds tests and fixes the issue so we don't end up with urls with too many or not enough slashes depending on how these settings are configured.

PragTob commented on lib/wallaby/browser.ex:
```diff
@@ -937,7 +937,10 @@ defmodule Wallaby.Browser do
   defp request_url(path) do
-    base_url() <> path
+    base_url = base_url() |> String.trim_trailing("/")
+    path = path |> String.trim_leading("/")
```
I think style wise piping when there's only one function call in the pipe is discouraged aka maybe rather `String.trim_leading(path, "/")` ?

aaronrenner replied:
Good call. I'll update the code.

PragTob commented on test/wallaby/browser_test.exs:
```diff
@@ -1,7 +1,100 @@
+    def visit(%Session{}, path) do
+      Agent.get(__MODULE__, fn (test_pid) ->
+        send test_pid, {:visit, path}
+      end)
```
:+1:

PragTob commented on test/wallaby/browser_test.exs:
```diff
@@ -1,7 +1,100 @@
+    test "relative path with leading slash, base url no trailing slash" do
+      Application.put_env(:wallaby, :base_url, "http://example.com")
+
+      session = %Session{driver: EchoingDriver}
```
couldn't we also create the session in the setup and pass it into the tests? Seems to be needed everywhere, no biggy though :)

PragTob commented on test/wallaby/browser_test.exs:
```diff
@@ -1,7 +1,100 @@
+      end
+    end
+
+    test "absolute url when the base_url isn't configured" do
```
good test :+1:

aaronrenner commented:
@PragTob I just updated `Browser.request_path/1` to not use pipes like you suggested. I also renamed `EchoingDriver` to `TestDriver` due to some further refactoring I did on the tests.

I thought about creating a session in the setup block, but then it wouldn't be immediately apparent that we're using a test driver instead of one of the real drivers. I agree defining the struct inline is not very pretty, so I extracted it into a helper.

```elixir
session = session_for_driver(TestDriver)
```

After reading the tests again this morning, when I saw `assert_received {:visit, url}` it wasn't immediately clear why the test process is receiving this message. To figure it out, I would have to go to the old `EchoingDriver` to see this message actually being sent.

Instead of having an `EchoingDriver` that sent the test process messages, I decided to try a custom assertion instead. Also, since the driver wasn't echoing messages anymore, I renamed it to `TestDriver`:

```elixir
TestDriver.assert_visited(url)
```

Now, the test driver keeps a list of all of the urls we've visited in this session and is able to raise a custom `ExUnit.AssertionError` when we don't visit the expected url.

Please let me know what you think and if this approach is clearer than what I had before.

PragTob commented on test/wallaby/browser_test.exs:
```diff
@@ -1,7 +1,122 @@
+  defmodule TestDriver do
+
+    def start_link() do
+      Agent.start_link(fn -> [] end, name: __MODULE__)
```
question about this, we register it under the current module name - so this disallows us from running these test async, correct? Couldn't we just start one and embed the information of the pid of the agent in the `TestDriver`/Session struct so that we could still run them in parallel and not rely on the name?

Alternatively, we could rely on module name + current test description (describe + test) if we can get it, should also be unique.

aaronrenner replied:
Unfortunately we can't run this test case with `async: true` because we're modifying the environment with `Application.put_env(:wallaby, :base_url, "http://example.com")` in some of the tests. 

If that wasn't the case, I agree it would be a great idea to see if we could make this driver not rely on a single registered name. If we reuse this `TestDriver` pattern somewhere else, I'd love to get it to work async.

PragTob replied:
Ah damn, totally I knew I probably overlooked something :D I mean, unless we start putting the base_url in the session right at startup or something like that, but that's a discussion for another time

PragTob commented on test/wallaby/browser_test.exs:
```diff
@@ -1,7 +1,122 @@
+      session = session_for_driver(TestDriver)
+      Browser.visit(session, "form.html")
+
+      TestDriver.assert_visited("http://example.com/form.html")
```
I like putting the assertion in the driver, no implementation leakage that way! :+1:
